src/herder/Upgrades.cpp

```diff
@@ -318,4 +328,237 @@ Upgrades::timeForUpgrade(uint64_t time) const
+{
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        return balanceAboveReserve;
+    }
+
+    if (accountID == getIssuer(asset))
+    {
+        return INT64_MAX;
+    }
+    else
+    {
+        auto trust = TrustFrame::loadTrustLine(accountID, asset, db, ld);
+        assert(trust);
+        return trust->getBalance();
+    }
+}
+
+static int64_t
+getAvailableLimit(AccountID const& accountID, Asset const& asset,
+                  int64_t balanceAboveReserve, LedgerDelta& ld, Database& db)
+{
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        return INT64_MAX - balanceAboveReserve;
+    }
+
+    if (accountID == getIssuer(asset))
+    {
+        return INT64_MAX;
+    }
+    else
+    {
+        auto trust = TrustFrame::loadTrustLine(accountID, asset, db, ld);
+        assert(trust);
+        return trust->getTrustLine().limit - trust->getBalance();
+    }
+}
+
+static bool
+shouldDeleteOffer(AccountID const& accountID, Asset const& asset,
+                  int64_t balanceAboveReserve,
+                  std::map<Asset, std::unique_ptr<int64_t>> const& liabilities,
+                  std::function<int64_t(AccountID const&, Asset const&, int64_t,
+                                        LedgerDelta&, Database&)>
+                      getCap,
+                  LedgerDelta& ld, Database& db)
+{
+    auto iter = liabilities.find(asset);
+    assert(iter != liabilities.end());
+    return iter->second ? *iter->second > getCap(accountID, asset,
+                                                 balanceAboveReserve, ld, db)
+                        : true;
+}
+
+static void
+prepareLiabilities(LedgerManager& ledgerManager, LedgerDelta& ld)
+{
+    auto& db = ledgerManager.getDatabase();
+    db.getEntryCache().clear();
+    auto offersByAccount = OfferFrame::loadAllOffers(db);
+
+    for (auto& accountOffers : offersByAccount)
+    {
+        std::map<Asset, std::unique_ptr<int64_t>> initialBuyingLiabilities;
+        std::map<Asset, std::unique_ptr<int64_t>> initialSellingLiabilities;
+        for (auto const& offerFrame : accountOffers.second)
+        {
+            auto const& offer = offerFrame->getOffer();
+            addLiabilities(initialBuyingLiabilities, offer.sellerID,
+                           offer.buying, offerFrame->getBuyingLiabilities());
+            addLiabilities(initialSellingLiabilities, offer.sellerID,
+                           offer.selling, offerFrame->getSellingLiabilities());
+        }
+
+        auto accountFrame =
+            AccountFrame::loadAccount(ld, accountOffers.first, db);
+        auto& account = accountFrame->getAccount();
+        assert(accountFrame);
+        int64_t balanceAboveReserve =
+            accountFrame->getBalanceAboveReserve(ledgerManager);
+
+        std::map<Asset, Liabilities> liabilities;
+        for (auto const& offerFrame : accountOffers.second)
+        {
+            auto const& offer = offerFrame->getOffer();
+            bool erase = shouldDeleteOffer(
```

Comment:
It would be cleaner to use `std::bind` with placeholders instead of having a bunch of parameters passed in to `shouldDeleteOffer` for the sole purpose of the `getBalance` function.
(better would have been to use lambdas and captures, but that function is already big enough)